engine-types/src/types.rs

```diff
@@ -1,14 +1,129 @@
-use super::{str, vec, Add, Address, String, Sub, Vec, U256};
+use crate::{str, vec, Address, AsMut, AsRef, Deref, DerefMut, Display, String, Vec, U256};
 use borsh::{BorshDeserialize, BorshSerialize};
 
+use crate::fmt::Formatter;
+pub use evm_core::Opcode;
+
 pub type Balance = u128;
 pub type RawAddress = [u8; 20];
 pub type RawU256 = [u8; 32]; // Big-endian large integer type.
 pub type RawH256 = [u8; 32]; // Unformatted binary data of fixed length.
 pub type EthAddress = [u8; 20];
-pub type Gas = u64;
 pub type StorageUsage = u64;
 
+#[derive(Default, Debug, Clone, Copy, Eq, PartialEq, Ord, PartialOrd)]
+/// Near gas type which wraps an underlying u64.
+pub struct NearGas(u64);
+
+impl Display for NearGas {
+    fn fmt(&self, f: &mut Formatter<'_>) -> crate::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl Deref for NearGas {
```

Comment:
I thought we agreed we do not want to implement `Deref` for newtypes.

Reply:
The automagic `Deref` implementation from the `derive_more` crate added some auto magic that wasn't intended. Manually, it doesn't have that negative issues previously experienced.

Reply:
I still don't think it's needed. Just gives another way to subvert the type system. If a function requires the raw type, but we have the gas type then we should ask ourselves why that is the case. If there is an easy escape hatch then we may be lazy and not ask this question when we should.

Of course it should possible to obtain the raw type because sometimes that will be needed (namely at the boundary of our application). But in those cases we should be clear about this (for example `fn raw(self) -> u64`).

Reply:
Fair point, double checked other newtype patterns, and correct, we shouldn't be adding `Deref` in this case, especially with what you had stated. Gas throughout precompiles does need to have some math done on it with other types.